chain/rosetta-rpc/src/adapters/mod.rs

```diff
@@ -144,54 +142,66 @@ pub(crate) async fn convert_block_to_transactions(
-                receipt_hash,
-            } => format!("receipt:{}", receipt_hash.to_base()),
-            near_primitives::views::StateChangeCauseView::ActionReceiptGasReward {
-                receipt_hash,
-            } => format!("receipt:{}", receipt_hash.to_base()),
-            near_primitives::views::StateChangeCauseView::ReceiptProcessing { receipt_hash } => {
+            StateChangeCauseView::ActionReceiptProcessingStarted { receipt_hash }
+            | StateChangeCauseView::ActionReceiptGasReward { receipt_hash }
+            | StateChangeCauseView::ReceiptProcessing { receipt_hash }
+            | StateChangeCauseView::PostponedReceipt { receipt_hash } => {
                 format!("receipt:{}", receipt_hash.to_base())
             }
-            near_primitives::views::StateChangeCauseView::PostponedReceipt { receipt_hash } => {
-                format!("receipt:{}", receipt_hash.to_base())
+            StateChangeCauseView::InitialState => format!("block:{}", block_hash),
+            StateChangeCauseView::ValidatorAccountsUpdate => {
+                format!("block-validators-update:{}", block_hash)
             }
-            near_primitives::views::StateChangeCauseView::InitialState
-            | near_primitives::views::StateChangeCauseView::ValidatorAccountsUpdate
-            | near_primitives::views::StateChangeCauseView::UpdatedDelayedReceipts => {
-                format!("block:{}", block.header.hash)
+            StateChangeCauseView::UpdatedDelayedReceipts => {
+                format!("block-delayed-receipts:{}", block_hash)
             }
-            near_primitives::views::StateChangeCauseView::NotWritableToDisk => unreachable!(),
-        };
-        let current_transaction = if let Some(transaction) = transactions.last_mut() {
-            if transaction.transaction_identifier.hash == transaction_hash {
-                Some(transaction)
-            } else {
-                None
+            StateChangeCauseView::NotWritableToDisk => {
+                return Err(crate::errors::ErrorKind::InternalInvariantError(
+                    "State Change 'NotWritableToDisk' should never be observed".to_string(),
+                ));
             }
-        } else {
-            None
         };
-        let current_transaction = if let Some(transaction) = current_transaction {
-            transaction
-        } else {
-            transactions.push(crate::models::Transaction {
-                transaction_identifier: crate::models::TransactionIdentifier {
-                    hash: transaction_hash.clone(),
-                },
-                operations: vec![],
-                metadata: crate::models::TransactionMetadata {
-                    type_: crate::models::TransactionType::Transaction,
-                },
+
+        let current_transaction =
+            transactions.entry(transaction_hash.clone()).or_insert_with(move || {
```

Comment:
Instead of relying on invalid sequential order (the state changes are grouped by account ID instead of preserving in historical order), I switched to using HashMap. If the order of transactions is critical, we should replicate the sorting after we collect this HashMap.